sycl/doc/EnvironmentVariables.md

```diff
@@ -12,8 +12,9 @@ subject to change. Do not rely on these variables in production code.
 | Environment variable | Values | Description |
 | -------------------- | ------ | ----------- |
 | SYCL_PI_TRACE | Described [below](#sycl_pi_trace-options)  | Enable specified level of tracing for PI. |
-| SYCL_BE | PI_OPENCL, PI_LEVEL_ZERO, PI_CUDA | Force SYCL RT to consider only devices of the specified backend during the device selection. |
-| SYCL_DEVICE_TYPE | One of: CPU, GPU, ACC, HOST | Force SYCL to use the specified device type. If unset, default selection rules are applied. If set to any unlisted value, this control has no effect. If the requested device type is not found, a `cl::sycl::runtime_error` exception is thrown. If a non-default device selector is used, a device must satisfy both the selector and this control to be chosen. This control only has effect on devices created with a selector. |
+| SYCL_BE | PI_OPENCL, PI_LEVEL_ZERO, PI_CUDA | Force SYCL RT to consider only devices of the specified backend during the device selection. We are planning to deprecate SYCL_BE environment variable in the future. The specific grace period is not decided yet. Please use the new env var SYCL_DEVICE_FILTER instead. |
+| SYCL_DEVICE_TYPE | One of: CPU, GPU, ACC, HOST | Force SYCL to use the specified device type. If unset, default selection rules are applied. If set to any unlisted value, this control has no effect. If the requested device type is not found, a `cl::sycl::runtime_error` exception is thrown. If a non-default device selector is used, a device must satisfy both the selector and this control to be chosen. This control only has effect on devices created with a selector. We are planning to deprecate SYCL_DEVICE_TYPE environment variable in the future. The specific grace period is not decided yet. Please use the new env var SYCL_DEVICE_FILTER instead. |
+| SYCL_DEVICE_FILTER (tentative name) | {backend:device_type:device_num} | Force SYCL RT to use the specified device with optional backend and device_num information. It can list any number of triples separated by commas. This env var affects all different device selectors and device discovery. When no device that can satisfy the triple by a default_selector, a heuristic will choose the device that has the closest match. HOST device is always available for the default_selector and the filter won't affect HOST device availability in any case. Possible values of device_type are \*,host,cpu,gpu,acc. Possible values of backend are \*,opencl, level_zero, cuda. Device_num is an unsigned integer that indexes the enumeration of devices from sycl::platform::get_device() call. All triple entries are optional, but one of them should be present. For example, to use cpu and level_zero gpu device number 0, one can set SYCL_DEVICE_FILTER=cpu,level_zero:gpu::0. This environment variable will limit loading only specified plugins into the SYCL RT. For example, SYCL_DEVICE_FILTER=level_zero and cpu_selector will throw an error because the CPU device is not supported by the level_zero backend. |
```

Comment:
I'd like to propose an updated wording for this description:

> Limits the SYCL RT to use only a subset of the system's devices.  Setting this environment variable affects all of the device query functions and all of the device selectors.  The value of this environment variable is a comma separated list of filters, where each filter is a triple of the form "backend:device_type:device_num" (without the quotes).  Each element of the triple is optional, but each filter must have at least one value.  Possible values of "backend" are "level_zero", "opencl", "cuda", or "\*".  Possible values of "device_type" are "cpu", "gpu", "acc", or "\*".  Device_num is an integer that indexes the enumeration of devices from the sycl::platform::get_device() call, where the first device in that enumeration has index zero.  Assuming a filter has all three elements of the triple, it selects only those devices that come from the given backend, have the specified device type, AND have the given device index.  If more than one filter is specified, the RT is restricted to the union of devices selected by all filters.  The filter does not allow the host device to be selected since the SYCL language requires this device to be always available.
>
> Note that the standard selectors like gpu_selector or cpu_selector will throw an exception if the filtered list of devices does not include a device that satisfies the selector.  In particular, limiting the devices to only those supported by the "level_zero" backend will cause the cpu_selector to throw an exception since that backend does not support any CPU devices.

I intentionally removed "host" as a possible value for "device_type" since the host device is unaffected by this environment variable.  I also intentionally left out any mention of loading plugins, since that is an implementation detail.

I added the second paragraph only because previous comments indicated that users were confused about the effect on standard selectors.  I'd be OK removing that paragraph if I'm wrong about that.

Reply:
This version looks better IMO. @jbrodman can you check please everything is accurate?

Reply:
> I'd like to propose an updated wording for this description:
> 
> > Limits the SYCL RT to use only a subset of the system's devices.  Setting this environment variable affects all of the device query functions and all of the device selectors.  The value of this environment variable is a comma separated list of filters, where each filter is a triple of the form "backend:device_type:device_num" (without the quotes).  Each element of the triple is optional, but each filter must have at least one value.  Possible values of "backend" are "level_zero", "opencl", "cuda", or "*".  Possible values of "device_type" are "cpu", "gpu", "acc", or "*".  Device_num is an integer that indexes the enumeration of devices from the sycl::platform::get_device() call, where the first device in that enumeration has index zero.  Assuming a filter has all three elements of the triple, it selects only those devices that come from the given backend, have the specified device type, AND have the given device index.  If more than one filter is specified, the RT is restricted to the union of devices selected by all filters.  The filter does not allow the host device to be selected since the SYCL language requires this device to be always available.
> > Note that the standard selectors like gpu_selector or cpu_selector will throw an exception if the filtered list of devices does not include a device that satisfies the selector.  In particular, limiting the devices to only those supported by the "level_zero" backend will cause the cpu_selector to throw an exception since that backend does not support any CPU devices.
> 
> I intentionally removed "host" as a possible value for "device_type" since the host device is unaffected by this environment variable. I also intentionally left out any mention of loading plugins, since that is an implementation detail.
> 
> I added the second paragraph only because previous comments indicated that users were confused about the effect on standard selectors. I'd be OK removing that paragraph if I'm wrong about that.

Thanks for your suggestion, @gmlueck. We discussed the rationale of possible values of HOST device_type, and we decided it is better to include it because leaving HOST out may give a wrong impression that HOST is not supported. Further, SYCL_DEVICE_FILTER=host will allow only HOST device to be used, which is a legal usage of the filter. 
Regarding the plugin loading statement, we also discussed about this. The new env var is truly a multi purposed. We originally thought about having 2 separate env vars to control plugin loading and to allow designation of desired devices. In our plugin spec, we mentioned this user plugin control logic. Later, we decided to use a single env var for both purposes. It is important to specify two different purposes to the users.

Reply:
> SYCL_DEVICE_FILTER=host will allow only HOST device to be used, which is a legal usage of the filter

I can see this argument.  If we don't allow "host", then the user needs to say "SYCL_DEVICE_FILTER=" in order to exclude all devices other than the host, and that does seem weird.  Note, however, that "host" is in the "backend" position of the filter, not the "device_type" position.  This is correct, since the host device exists in its own platform (thus its own backend).    In fact, it's not really useful to allow "host" as a device_type since the host device exists only in the host backend (and the host backend has no other device types).  However, it seems reasonable to include it as a device type anyway.

Here's an updated proposal:

> Limits the SYCL RT to use only a subset of the system's devices. Setting this environment variable affects all of the device query functions and all of the device selectors. The value of this environment variable is a comma separated list of filters, where each filter is a triple of the form "backend:device_type:device_num" (without the quotes). Each element of the triple is optional, but each filter must have at least one value. Possible values of "backend" are "host", "level_zero", "opencl", "cuda", or "\*". Possible values of "device_type" are "host", "cpu", "gpu", "acc", or "*". Device_num is an integer that indexes the enumeration of devices from the sycl::platform::get_device() call, where the first device in that enumeration has index zero. Assuming a filter has all three elements of the triple, it selects only those devices that come from the given backend, have the specified device type, AND have the given device index. If more than one filter is specified, the RT is restricted to the union of devices selected by all filters. The RT always includes the "host" backend and the host device regardless of the filter because the SYCL language requires this device to always be present.  Therefore, including "host" in the list of filters is allowed but is unnecessary.
>
> Note that the standard selectors like gpu_selector or cpu_selector will throw an exception if the filtered list of devices does not include a device that satisfies the selector. In particular, limiting the devices to only those supported by the "level_zero" backend will cause the cpu_selector to throw an exception since that backend does not support any CPU devices.

I don't think I agree about including information about plugin loading in the description.  This seems like an implementation detail to me.  How many of our customers even know what a DPC++ plugin is?  Do we allow them to add new plugins?  (I didn't think so.)  If plugins are not otherwise exposed to our customers, it is wrong to include them in the documentation for this env var.

Reply:
> Note, however, that "host" is in the "backend" position of the filter, not the "device_type" position.

Yes, HOST is included in the backend type, but HOST is also one of SYCL device_types we support. https://github.com/intel/llvm/blob/f253851f7ab67b5704f298dd8ec7695a11b64706/sycl/include/CL/sycl/info/info_desc.hpp#L143

> How many of our customers even know what a DPC++ plugin is? Do we allow them to add new plugins? (I didn't think so.) If plugins are not otherwise exposed to our customers, it is wrong to include them in the documentation for this env var.

I cannot answer how many of them, but our customers are aware of the plugins.
 t is an important part of SYCL to use it correctly.

https://github.com/intel/llvm/blob/sycl/sycl/doc/FAQ.md#q-what-devices-are-supported-by-dpc-compiler
https://github.com/intel/llvm/blob/sycl/sycl/doc/PluginInterface.md
Our customers have been using SYCL_BE=PI_LEVEL_ZERO to set the gpu backend plugin.
Our customers uses SYCL_PI_TRACE if they want to see details of plugin calls.

As written in the plugin spec, PIs are designed to allow users to add / remove plugins.
Currently, SYCL_DEVICE_FILTER is the only way they can control of loading plugins.

Reply:
> Yes, HOST is included in the backend type, but HOST is also one of SYCL device_types we support.

I think there is no disagreement.  I included "host" has both a backend and a device_type in my updated description.

> t is an important part of SYCL to use it correctly.

I'm still not convinced :-)  I think the description about plugin loading is not helpful in the documentation, and we should leave it out.  Here are some responses to your points:

* Users definitely don't need to know about plugins to use SYCL correctly.  The SYCL spec and SYCL API don't even mention plugins.

* The description of the existing `SYCL_BE` env var never mentioned the word "plugin".  It only said it restricted the set of **backends** that the RT uses.

* I agree that `SYCL_PI_TRACE` is specific to plugins, but I think this is used only by the developers of DPC++ itself, not by people using DPC++.

* The "PluginInterface.md" document seems to be more of an internal design document at this point, not a document that end users of DPC++ will use.  Perhaps it will be possible for end-users to add plugins in the future, but there are too many TBD's in that document today for this to be a reality.

Reply:
> I think the description about plugin loading is not helpful in the documentation

@gmlueck, I think with introducing SYCL_DEVICE_FILTER and SYCL_BE to end-users we are stepping one level down from 
what's specified by SYCL - device selectors. So to properly describe what SYCL_DEVICE_FILTER really does, plugin/plugin loading seems like a quite relevant implementation detail information.

Reply:
@bso-intel already added this sentence at the end of the description for `SYCL_DEVICE_FILTER`:

> This environment variable can be used to limit loading only specified plugins into the SYCL RT.

I'm not opposed to that, and I assume this addresses your concern @kbobrovs?

FWIW, I still think that few end users will know what a DPC++ plugin is, and this sentence will probably not make any sense to them.  However, the sentence is at the end, so most readers will just ignore it.